### DISCUSSION TITLE
Golang Unit tests

### DISCUSSION THREAD

**Discussion by sorsbyl** [paketo-buildpacks/feedback#63] [General]
I've been reading through the docs and I can't see a way to run custom commands for the golang buildpack.
In the nodejs buildpack you can specify specific commands to be run i.e `BP_NODE_RUN_SCRIPTS="test,lint"` which would allow us to fail the build if the unit tests failed.
I can't see anything in the golang builder that would allow us to do this. Surely, as part of the build we should be checking that things like tests pass?

**Comment by ryanmoran** [comment]
Buildpacks, at least those currently provided by Paketo, aren't designed to replace your entire CI/CD system. Their primary focus is to build a production-grade container image. Given that, we'd expect the code to have been tested prior to the point at which buildpacks would create this image.

If I think about it as being conceptually analogous to using a `Dockerfile` to create your image, is running your tests inside the `Dockerfile` a common pattern?

**Reply by sorsbyl** [reply]
Thanks for the reply @ryanmoran,
I've looked at a few pros and cons to this and there are a number of opinions around this topic.

This is my opinion:
A production-grade container image enforces that tests are performed to ensure quality and stability. If we was talking about just a plain old single layer Dockerfile then no I wouldn't as having test resources / sources / dependencies in our production grade containers isn't great practice.

However, I view buildpacks closer to multi-staged Dockerfiles rather than single layer Dockerfiles. This is common practice as documented in the docker documentation https://docs.docker.com/language/java/run-tests/#multi-stage-dockerfile-for-testing. (copying contents and posting below for ease of reading and to avoid jumping links)

```Dockerfile
FROM eclipse-temurin:17-jdk-jammy as base
WORKDIR /app
COPY .mvn/ .mvn
COPY mvnw pom.xml ./
RUN ./mvnw dependency:resolve
COPY src ./src

FROM base as test
CMD ["./mvnw", "test"]

FROM base as development
CMD ["./mvnw", "spring-boot:run", "-Dspring-boot.run.profiles=mysql", "-Dspring-boot.run.jvmArguments='-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:8000'"]

FROM base as build
RUN ./mvnw package


FROM eclipse-temurin:17-jre-jammy as production
EXPOSE 8080
COPY --from=build /app/target/spring-petclinic-*.jar /spring-petclinic.jar
CMD ["java", "-Djava.security.egd=file:/dev/./urandom", "-jar", "/spring-petclinic.jar"]
```

I would personally like to see us making use of tests as a separate task.

For languages such as nodejs. I'd expect to install dependencies twice, once with dev dependencies and another with just the production dependencies. The latter saved for use in production where as the former just cached for build/testing purposes.
